clash_lib/src/proxy/tuic/mod.rs

```diff
@@ -0,0 +1,296 @@
+    pub skip_cert_verify: bool,
+    pub sni: Option<String>,
+}
+
+pub struct Handler {
+    opts: HandlerOptions,
+    ep: TuicEndpoint,
+    conn: AsyncMutex<Option<TuicConnection>>,
+    next_assoc_id: AtomicU16
+}
+
+#[async_trait]
+impl OutboundHandler for Handler {
+    fn name(&self) -> &str {
+        &self.opts.name
+    }
+
+    fn proto(&self) -> OutboundType {
+        OutboundType::Tuic
+    }
+
+    async fn remote_addr(&self) -> Option<SocksAddr> {
+        None
+    }
+
+    async fn support_udp(&self) -> bool {
+        true
+    }
+
+    async fn proxy_stream(
+        &self,
+        s: AnyStream,
+        _sess: &Session,
+        _resolver: ThreadSafeDNSResolver,
+    ) -> std::io::Result<AnyStream> {
+        tracing::warn!("Proxy stream currently is direcrt connect");
+        Ok(s)
+    }
+
+    async fn connect_stream(
+        &self,
+        sess: &Session,
+        resolver: ThreadSafeDNSResolver,
+    ) -> std::io::Result<BoxedChainedStream> {
+        self.do_connect_stream(sess, resolver).await.map_err(|e| {
+            tracing::error!("{:?}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
+        })
+    }
+    async fn connect_datagram(
+        &self,
+        sess: &Session,
+        resolver: ThreadSafeDNSResolver,
+    ) -> std::io::Result<BoxedChainedDatagram> {
+        self.do_connect_datagram(sess, resolver).await.map_err(|e| {
+            tracing::error!("{:?}", e);
+            std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
+        })
+    }
+}
+
+impl Handler {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(opts: HandlerOptions) -> Result<AnyOutboundHandler, crate::Error> {
+        let mut crypto = TlsConfig::builder()
+            .with_safe_default_cipher_suites()
+            .with_safe_default_kx_groups()
+            .with_protocol_versions(&[&rustls::version::TLS13])
+            .unwrap()
+            .with_root_certificates(GLOBAL_ROOT_STORE.clone())
+            .with_no_client_auth();
+        // aborted by peer: the cryptographic handshake failed: error 120: peer doesn't support any known protocol
+        crypto.alpn_protocols = opts.alpn.clone();
+        crypto.enable_early_data = true;
+        crypto.enable_sni = !opts.disable_sni;
+        let mut quinn_config = QuinnConfig::new(Arc::new(crypto));
+        let mut quinn_transport_config = QuinnTransportConfig::default();
+        quinn_transport_config
+            .max_concurrent_bidi_streams(opts.max_open_stream)
+            .max_concurrent_uni_streams(opts.max_open_stream)
+            .send_window(opts.send_window)
+            .stream_receive_window(opts.receive_window)
+            .max_idle_timeout(None)
+            .congestion_controller_factory(Arc::new(CubicConfig::default()));
+        quinn_config.transport_config(Arc::new(quinn_transport_config));
+        // Try to create an IPv4 socket as the placeholder first, if it fails, try IPv6.
+        let socket =
```

Comment:
would this work if server is v6 only and v4 is used locally?

Reply:
I won't think it will work. Should we check server dns record is A or AAAA?
And if server is daul-stack, does clash-rs supply options prefer IPv4 or IPv6?

Reply:
i think we should check.

if you use clash resolver, it will try v6 first https://github.com/Watfaq/clash-rs/blob/master/clash_lib/src/app/dns/resolver.rs#L423

Reply:
I'm approving this for now. feel free to check in and make a separate PR for this as well as other features mentioned in the PR desc.

thanks!